subsys/fs/zffs/src/zffs_dlist.c

```diff
@@ -0,0 +1,252 @@
+{
+	int rc;
+	struct dlist_node_disk_head head;
+	struct dlist_node_disk_tail tail;
+	u16_t crc;
+
+	rc = zffs_object_new(zffs, pointer, node->id,
+			      ex_len + sizeof(head) + sizeof(tail));
+	if (rc) {
+		return rc;
+	}
+
+	head.object_type = ZFFS_OBJECT_TYPE_DLIST_NODE;
+	sys_put_le32(node->next, head.next);
+	sys_put_le32(node->prev, head.prev);
+	crc = crc16_ccitt(0, (const void *)&head, sizeof(head));
+	rc = zffs_area_write(zffs, pointer, &head, sizeof(head));
+	if (rc) {
+		return rc;
+	}
+
+	if (ex_len) {
+		crc = crc16_ccitt(crc, ex_data, ex_len);
+		rc = zffs_area_write(zffs, pointer, ex_data, ex_len);
+		if (rc) {
+			return rc;
+		}
+	}
+
+	sys_put_le16(crc, tail.crc);
+	return zffs_area_write(zffs, pointer, (const void *)&tail,
+				sizeof(tail));
+}
+
+static int dlist_node_update(struct zffs_data *zffs,
+			     struct zffs_area_pointer *to, u32_t id,
+			     u32_t *prev, u32_t *next, const void *ex_data,
+			     u32_t update_ex_len)
+{
+	int rc;
+	struct zffs_area_pointer from = *to;
+	struct dlist_node_disk_head head;
+	struct dlist_node_disk_tail tail;
+	u16_t crc;
+	u32_t ex_len;
+
+	rc = zffs_object_open(zffs, &from, id, NULL);
+	if (rc < 0) {
+		return rc;
+	}
+
+	if (rc < sizeof(head) + sizeof(tail)) {
+		return -EIO;
+	}
+
+	ex_len = rc - sizeof(head) - sizeof(tail);
+	if (!ex_data) {
+		update_ex_len = 0;
+	}
+
+	rc =
+	    zffs_object_update(zffs, to, id, max(ex_len, update_ex_len) +
+						   sizeof(head) + sizeof(tail));
+	if (rc) {
+		return rc;
+	}
+
+	rc = zffs_area_read(zffs, &from, &head, sizeof(head));
+	if (rc) {
+		return rc;
+	}
+
+	if (head.object_type != ZFFS_OBJECT_TYPE_DLIST_NODE) {
+		return -EIO;
+	}
+
+	if (next) {
+		sys_put_le32(*next, head.next);
+	}
+
+	if (prev) {
+		sys_put_le32(*prev, head.prev);
+	}
+
+	crc = crc16_ccitt(0, (const void *)&head, sizeof(head));
+	rc = zffs_area_write(zffs, to, &head, sizeof(head));
+	if (rc) {
+		return rc;
+	}
+
+	if (ex_data) {
+		rc = zffs_area_write(zffs, to, ex_data, update_ex_len);
+		if (rc) {
+			return rc;
+		}
+		crc = crc16_ccitt(crc, ex_data, update_ex_len);
+
+		from.offset += update_ex_len;
+	}
+
+	if (update_ex_len < ex_len) {
+		rc = zffs_area_copy_crc(zffs, &from, to,
+					 ex_len - update_ex_len, &crc);
+		if (rc) {
+			return rc;
+		}
+	}
+
+	sys_put_le16(crc, tail.crc);
+	return zffs_area_write(zffs, to, (const void *)&tail, sizeof(tail));
+}
+
+static int dlist_node_load(struct zffs_data *zffs,
+			   struct zffs_area_pointer *pointer,
+			   struct zffs_dlist_node *node)
+{
+	int rc;
+	u32_t ex_len;
+	struct dlist_node_disk_head head;
+
+	rc = zffs_object_open(zffs, pointer, node->id, NULL);
+	if (rc < 0) {
+		return rc;
+	}
+
+	if (rc < sizeof(head) + sizeof(struct dlist_node_disk_tail)) {
+		return -EIO;
+	}
+
+	ex_len = rc - sizeof(head) + sizeof(struct dlist_node_disk_tail);
+
+	rc = zffs_area_read(zffs, pointer, &head, sizeof(head));
+	if (rc) {
+		return rc;
+	}
+	if (head.object_type != ZFFS_OBJECT_TYPE_DLIST_NODE) {
```

Comment:
have an empty line after a } (unless it's an else or another }) Apply that everywhere where relevant

Reply:
yep